### DISCUSSION TITLE
Colorization instance map question

### DISCUSSION THREAD

**Discussion by GrimmSnark** [neuropoly/axondeepseg#841] [Help needed]
Hello,

So I have been playing around with the colorization feature in the morphometrics extraction pipeline. My question refers to the colorization instance map and how that relates to the morphometrics extraction. The segmentation I have with my images is pretty good at the moment. However the colorization instance map shows myelin identity boundary creep between touching axons. 

Does the colorization identity mismatch contribute to the calculation of the myelin thickness? And if this is the case what is the best way to address this issue?

Thanks a lot,

Michael 

![LM_1](https://user-images.githubusercontent.com/10127351/191260286-6d22e776-0c74-4a47-97f1-215606b8b460.png)

![LM_1_axonmyelin_index](https://user-images.githubusercontent.com/10127351/191258756-c10617dc-661c-4602-b6c5-9dc2dd0e3b42.png)

![LM_1_instance-map](https://user-images.githubusercontent.com/10127351/191258823-21c1c3cd-cb56-4cf0-9b08-f3060c49d1de.png)

**Comment by hermancollin** [comment]
Thank you @GrimmSnark for reaching out again. The deep-learning models that ADS uses produce a **semantic segmentation** (which is the one in red and blue above). This image associates every pixel to one of the 3 classes: axon/myelin/background.
The instance segmentation, however, is obtained by an algorithm that uses the semantic segmentation to separate individual instances. This algorithm is responsible for what you call the *myelin identity boundary*. A well-known downfall of this method is that for **touching** instances, the algorithm has to chose where the boundary will be, and under certain geometric circumstances, the output is somewhat unpredictable.
As you might expect, this instance segmentation with "glitchy" myelin boundaries is the data used to extract metrics (because the metrics calculations are applied on individual axons, so we *need* this instance segmentation representation)

---

![animated](https://user-images.githubusercontent.com/83031821/191309049-2ac3d867-58c8-40f4-b17e-0011dd7a2db7.gif)

I cropped a part of your image to demonstrate how the metrics could be improved. Let's take axon #822 as an example. I see 2 options.

1. You could correct the segmentations manually. For example, axon #822 touches a few small axons. We can see that this is the reason why there is a "leak" of myelin from other instances inside the axon 822's myelin. By "erasing" the myelin inbetween, we will obtain the instance segmentation we are expecting. The correction would look like so:
![sema_corr](https://user-images.githubusercontent.com/83031821/191311273-e92791cd-b202-453d-b9b1-a999da0c78dd.png)
For this segmentation, the watershed algorithm will know that axon #822 is isolated from the other ones, and it will be a single instance instead of a single one with multiple "leaks" inside (i.e. if you look at your original image, notice how the axons that are not adjacent to any other axons are perfectly represented in the instance segmentation).

2. The solution 1 could be time-consuming. What I would try first if I were you would be to play around with the zoom factor (using the zoom factor sweep feature) to see if there is a zoom factor that produces less "adjacency" (i.e. I think your current segmentation is very neat but myelin looks a bit over-segmented). It is possible you could find a zoom factor that leads to easier manual correction (described in solution 1).

Keep in mind: these are issues we are well aware of and we are actively searching for solutions to this problem.

**Comment by GrimmSnark** [comment]
Hello @hermancollin,

Thank you again for your very quick and informative response! 

Do you think it might be a viable option to implement the option for a watershed segementation step (i.e. something like skimage.segmentation  watershed) to help correct the myelin mask before moving into the morphometrics extraction phase of the program?

I have also had some success in other projects with implementing an erode filter (i.e skimage.morphology erode) which removes a single pixel line from the entire circumference of the ROI mask. But this really only helps in very limited cases. 

As for the zoom factor, I will certainly try it, but do you have a suggestion on a starting range for that variable? 

N.B. Do you mind if I ask how you created the gif in the comment? Would be very useful for me going forward.

All the best,

Michael

**Comment by Stoyan-I-A** [comment]
> Do you think it might be a viable option to implement the option for a watershed segementation step (i.e. something like skimage.segmentation watershed) to help correct the myelin mask before moving into the morphometrics extraction phase of the program?

The instance segmentation actually uses the watershed from skimage, with its seed points being the centroids of the axons (blue region).

**Comment by hermancollin** [comment]
> As for the zoom factor, I will certainly try it, but do you have a suggestion on a starting range for that variable?
> 
> N.B. Do you mind if I ask how you created the gif in the comment? Would be very useful for me going forward.

I would recommend to start with a range from 0.90 to 1.10 with, say, 4 values (i.e. `l=4`). If you don't see enough difference, you can increase the range and the length gradually. That's a typical workflow for us.

As for the GIFs, I use this: https://ezgif.com/maker

**Comment by mathieuboudreau** [comment]
@GrimmSnark have you tried @hermancollin's suggestion?

**Comment by mathieuboudreau** [comment]
Converting this into the user discussions section
